File: backend/src/modules/siteHierarchy/siteHierarchy.service.js
/**
 * Canonical site/building/floor/equipment hierarchy from Prisma.
 * Used to sync engineering working-version payload with relational data.
 */

const prisma = require('../../lib/prisma');
const { sortBySortOrderThenName } = require('../../lib/hierarchySort');

/** Site Builder form labels (or API enum strings) → Prisma EntityStatus */
function entityStatusFromFormLabel(label) {
  const u = String(label || '')
    .trim()
    .toUpperCase();
  if (u === 'INACTIVE' || u === 'DRAFT') return 'INACTIVE';
  if (u === 'ARCHIVED') return 'ARCHIVED';
  if (u === 'ACTIVE') return 'ACTIVE';
  const lower = String(label || '')
    .trim()
    .toLowerCase();
  if (lower === 'draft' || lower === 'inactive') return 'INACTIVE';
  if (lower === 'archived') return 'ARCHIVED';
  if (lower === 'active') return 'ACTIVE';
  return 'ACTIVE';
}

/** Any client input → Prisma EntityStatus (handles title case, already-enums, etc.) */
function normalizeEntityStatusForDb(value) {
  if (value === undefined || value === null) return undefined;
  const u = String(value).trim().toUpperCase();
  if (u === 'ACTIVE' || u === 'INACTIVE' || u === 'ARCHIVED') return u;
  return entityStatusFromFormLabel(value);
}

/** Prisma EntityStatus → Site Builder form labels */
function formLabelFromEntityStatus(st) {
  const u = String(st || '').toUpperCase();
  if (u === 'INACTIVE') return 'Draft';
  if (u === 'ARCHIVED') return 'Archived';
  return 'Active';
}

function pointToWorkspaceRow(equipmentId, equipmentName, pt) {
  const units = pt.unit || '';
  const val = pt.presentValue != null && pt.presentValue !== '' ? String(pt.presentValue) : '—';
  const valueStr = units ? `${val} ${units}`.trim() : val;
  return {
    id: `${equipmentId}-${pt.pointCode || pt.id}`,
    databasePointId: pt.id,
    equipmentId,
    equipmentName,
    pointId: pt.pointCode,
    pointKey: pt.pointCode,
    pointDescription: pt.pointName,
    pointName: pt.pointName,
    pointReferenceId: pt.pointCode,
    value: valueStr,
    units,
    status: pt.status === 'ACTIVE' ? 'OK' : 'Warn',
    writable: pt.writable,
  };
}

/**
 * @returns {{ site: object, equipment: object[] }}
 */
async function buildWorkingSiteEquipmentFromDb(siteId) {
  const siteRow = await prisma.site.findUnique({ where: { id: siteId } });
  if (!siteRow) {
    return { site: null, equipment: [] };
  }

  const buildings = sortBySortOrderThenName(
    await prisma.building.findMany({
      where: { siteId },
      orderBy: { name: 'asc' },
    })
  );

  const siteBuildings = [];
  const allEquipment = [];

  for (const b of buildings) {
    const floors = sortBySortOrderThenName(
      await prisma.floor.findMany({
        where: { buildingId: b.id },
        orderBy: { name: 'asc' },
      })
    );
    const layoutStatus = b.status === 'ACTIVE' ? 'normal' : 'warning';
    const floorNodes = floors.map((f) => ({
      id: f.id,
      name: f.name,
      sortOrder: f.sortOrder ?? 0,
      floorType: f.floorType || 'Standard Floor',
      occupancyType: f.occupancyType || '',
    }));

    siteBuildings.push({
      id: b.id,
      name: b.name,
      /** Empty when unset so Site Builder dropdown matches an option (no fake "Building" value). */
      buildingType: b.buildingType != null && String(b.buildingType).trim() ? String(b.buildingType).trim() : '',
      buildingCode: b.buildingCode || '',
      description: b.description != null && String(b.description).trim() ? String(b.description).trim() : '',
      address: b.addressLine1,
      city: b.city,
      state: b.state,
      lat: b.latitude,
      lng: b.longitude,
      status: formLabelFromEntityStatus(b.status),
      layoutStatus,
      sortOrder: b.sortOrder ?? 0,
      hasFloors: floorNodes.length > 0,
      floors: floorNodes,
    });

    for (const f of floors) {
      const equipmentList = await prisma.equipment.findMany({
        where: { floorId: f.id },
        orderBy: { name: 'asc' },
        include: { controllersMapped: true },
      });
      for (const eq of equipmentList) {
        const points = await prisma.point.findMany({
          where: { equipmentId: eq.id },
          orderBy: { pointCode: 'asc' },
        });
        const livePoints = points.map((p) => pointToWorkspaceRow(eq.id, eq.name, p));
        const ec = eq.controllersMapped;
        const engStatus = ec
          ? 'CONTROLLER_ASSIGNED'
          : eq.status === 'ACTIVE'
            ? 'MISSING_CONTROLLER'
            : 'DRAFT';
        allEquipment.push({
          id: eq.id,
          floorId: eq.floorId,
          siteId: eq.siteId,
          buildingId: eq.buildingId,
          name: eq.name,
          displayLabel: eq.name,
          type: eq.equipmentType,
          instanceNumber: eq.instanceNumber ?? null,
          equipmentType: eq.equipmentType,
          address: eq.address || '',
          locationLabel: '',
          controllerRef: ec ? ec.controllerCode : null,
          deviceInstance: ec?.deviceInstance ?? null,
          protocol: ec ? ec.protocol || 'BACnet/IP' : 'API',
          templateName: eq.templateName ?? null,
          pointsDefined: points.length,
          status: engStatus,
          notes: '',
          livePoints,
        });
      }
    }
  }

  const workingSite = {
    id: siteRow.id,
    name: siteRow.name,
    mode: 'api',
    status: 'editing',
    /** Site Builder "Status" dropdown (maps to Site.status in DB) */
    nodeStatus: formLabelFromEntityStatus(siteRow.status),
    siteType: siteRow.siteType || '',
    timezone: siteRow.timezone || '',
    displayLabel: siteRow.displayLabel || siteRow.name,
    description: siteRow.description || '',
    engineeringNotes: siteRow.engineeringNotes || '',
    icon: siteRow.icon || '',
    buildings: siteBuildings,
  };

  return { site: workingSite, equipment: allEquipment };
}

module.exports = {
  buildWorkingSiteEquipmentFromDb,
  entityStatusFromFormLabel,
  formLabelFromEntityStatus,
  normalizeEntityStatusForDb,
};

File: server/src/main/java/com/legioncontrols/server/sites/SiteHierarchyService.java
package com.legioncontrols.server.sites;

import com.legioncontrols.server.common.JsonUtil;
import com.legioncontrols.server.jooq.generated.tables.records.BuildingRecord;
import com.legioncontrols.server.jooq.generated.tables.records.FloorRecord;
import com.legioncontrols.server.jooq.generated.tables.records.PointRecord;
import com.legioncontrols.server.jooq.generated.tables.records.SiteRecord;
import java.util.Objects;
import org.jooq.DSLContext;
import org.jooq.Record;
import org.jooq.Result;
import org.springframework.stereotype.Service;
import tools.jackson.databind.node.ArrayNode;
import tools.jackson.databind.node.ObjectNode;

import static com.legioncontrols.server.jooq.generated.tables.Building.BUILDING;
import static com.legioncontrols.server.jooq.generated.tables.Controllersmapped.CONTROLLERSMAPPED;
import static com.legioncontrols.server.jooq.generated.tables.Equipment.EQUIPMENT;
import static com.legioncontrols.server.jooq.generated.tables.Floor.FLOOR;
import static com.legioncontrols.server.jooq.generated.tables.Point.POINT;
import static com.legioncontrols.server.jooq.generated.tables.Site.SITE;

/**
 * Java equivalent of backend/src/modules/siteHierarchy/siteHierarchy.service.js's
 * buildWorkingSiteEquipmentFromDb — the canonical Site/Building/Floor/Equipment/Point hierarchy
 * read from Postgres, used to sync the Engineering working-version payload with relational truth
 * and to populate a release's operator-facing site/equipment snapshot at build time.
 */
@Service
public class SiteHierarchyService {

    private final DSLContext dsl;
    private final JsonUtil json;

    public SiteHierarchyService(DSLContext dsl, JsonUtil json) {
        this.dsl = dsl;
        this.json = json;
    }

    private static String formLabelFromEntityStatus(String status) {
        if ("INACTIVE".equals(status)) return "Draft";
        if ("ARCHIVED".equals(status)) return "Archived";
        return "Active";
    }

    public record HierarchyResult(ObjectNode site, ArrayNode equipment) {
    }

    public HierarchyResult buildWorkingSiteEquipmentFromDb(String siteId) {
        SiteRecord siteRow = dsl.selectFrom(SITE).where(SITE.ID.eq(siteId)).fetchOne();
        ArrayNode equipmentOut = json.newArray();
        if (siteRow == null) {
            return new HierarchyResult(null, equipmentOut);
        }

        ArrayNode buildingsOut = json.newArray();

        Result<BuildingRecord> buildings = dsl.selectFrom(BUILDING)
            .where(BUILDING.SITEID.eq(siteId))
            .orderBy(BUILDING.SORTORDER.asc(), BUILDING.NAME.asc())
            .fetch();

        for (BuildingRecord b : buildings) {
            String buildingId = b.get(BUILDING.ID);
            String buildingStatus = b.get(BUILDING.STATUS).getLiteral();
            String layoutStatus = "ACTIVE".equals(buildingStatus) ? "normal" : "warning";

            Result<FloorRecord> floors = dsl.selectFrom(FLOOR)
                .where(FLOOR.BUILDINGID.eq(buildingId))
                .orderBy(FLOOR.SORTORDER.asc(), FLOOR.NAME.asc())
                .fetch();

            ArrayNode floorsOut = json.newArray();
            for (FloorRecord f : floors) {
                ObjectNode floorNode = json.newObject();
                floorNode.put("id", f.get(FLOOR.ID));
                floorNode.put("name", f.get(FLOOR.NAME));
                String displayLabel = f.get(FLOOR.DISPLAYLABEL);
                floorNode.put("displayLabel", displayLabel != null && !displayLabel.isBlank() ? displayLabel : f.get(FLOOR.NAME));
                floorNode.put("sortOrder", Objects.requireNonNullElse(f.get(FLOOR.SORTORDER), 0));
                String floorType = f.get(FLOOR.FLOORTYPE);
                floorNode.put("floorType", floorType != null && !floorType.isBlank() ? floorType : "Standard Floor");
                String occupancyType = f.get(FLOOR.OCCUPANCYTYPE);
                floorNode.put("occupancyType", occupancyType != null ? occupancyType : "");
                floorsOut.add(floorNode);
            }

            ObjectNode buildingNode = json.newObject();
            buildingNode.put("id", buildingId);
            buildingNode.put("name", b.get(BUILDING.NAME));
            String buildingType = b.get(BUILDING.BUILDINGTYPE);
            buildingNode.put("buildingType", buildingType != null && !buildingType.isBlank() ? buildingType.trim() : "");
            String buildingCode = b.get(BUILDING.BUILDINGCODE);
            buildingNode.put("buildingCode", buildingCode != null ? buildingCode : "");
            String description = b.get(BUILDING.DESCRIPTION);
            buildingNode.put("description", description != null && !description.isBlank() ? description.trim() : "");
            buildingNode.put("address", b.get(BUILDING.ADDRESSLINE1));
            buildingNode.put("city", b.get(BUILDING.CITY));
            buildingNode.put("state", b.get(BUILDING.STATE));
            Double lat = b.get(BUILDING.LATITUDE);
            Double lng = b.get(BUILDING.LONGITUDE);
            if (lat != null) buildingNode.put("lat", lat); else buildingNode.putNull("lat");
            if (lng != null) buildingNode.put("lng", lng); else buildingNode.putNull("lng");
            buildingNode.put("status", formLabelFromEntityStatus(buildingStatus));
            buildingNode.put("layoutStatus", layoutStatus);
            buildingNode.put("sortOrder", Objects.requireNonNullElse(b.get(BUILDING.SORTORDER), 0));
            buildingNode.put("hasFloors", !floorsOut.isEmpty());
            buildingNode.set("floors", floorsOut);
            buildingsOut.add(buildingNode);

            for (FloorRecord f : floors) {
                String floorId = f.get(FLOOR.ID);
                Result<Record> equipmentRows = dsl.select(EQUIPMENT.fields())
                    .select(CONTROLLERSMAPPED.CONTROLLERCODE, CONTROLLERSMAPPED.DEVICEINSTANCE, CONTROLLERSMAPPED.PROTOCOL)
                    .from(EQUIPMENT)
                    .leftJoin(CONTROLLERSMAPPED).on(CONTROLLERSMAPPED.EQUIPMENTID.eq(EQUIPMENT.ID))
                    .where(EQUIPMENT.FLOORID.eq(floorId))
                    .orderBy(EQUIPMENT.NAME.asc())
                    .fetch();

                for (Record eq : equipmentRows) {
                    String equipmentId = eq.get(EQUIPMENT.ID);
                    Result<PointRecord> points = dsl.selectFrom(POINT)
                        .where(POINT.EQUIPMENTID.eq(equipmentId))
                        .orderBy(POINT.POINTCODE.asc())
                        .fetch();

                    ArrayNode livePoints = json.newArray();
                    for (PointRecord p : points) {
                        livePoints.add(pointToWorkspaceRow(equipmentId, eq.get(EQUIPMENT.NAME), p));
                    }

                    String controllerCode = eq.get(CONTROLLERSMAPPED.CONTROLLERCODE);
                    String eqStatus = eq.get(EQUIPMENT.STATUS).getLiteral();
                    String engStatus = controllerCode != null
                        ? "CONTROLLER_ASSIGNED"
                        : "ACTIVE".equals(eqStatus) ? "MISSING_CONTROLLER" : "DRAFT";

                    ObjectNode equipmentNode = json.newObject();
                    equipmentNode.put("id", equipmentId);
                    equipmentNode.put("floorId", eq.get(EQUIPMENT.FLOORID));
                    equipmentNode.put("siteId", eq.get(EQUIPMENT.SITEID));
                    equipmentNode.put("buildingId", eq.get(EQUIPMENT.BUILDINGID));
                    equipmentNode.put("name", eq.get(EQUIPMENT.NAME));
                    equipmentNode.put("displayLabel", eq.get(EQUIPMENT.NAME));
                    equipmentNode.put("type", eq.get(EQUIPMENT.EQUIPMENTTYPE));
                    String instanceNumber = eq.get(EQUIPMENT.INSTANCENUMBER);
                    if (instanceNumber != null) equipmentNode.put("instanceNumber", instanceNumber); else equipmentNode.putNull("instanceNumber");
                    equipmentNode.put("equipmentType", eq.get(EQUIPMENT.EQUIPMENTTYPE));
                    String address = eq.get(EQUIPMENT.ADDRESS);
                    equipmentNode.put("address", address != null ? address : "");
                    equipmentNode.put("locationLabel", "");
                    if (controllerCode != null) equipmentNode.put("controllerRef", controllerCode); else equipmentNode.putNull("controllerRef");
                    String deviceInstance = eq.get(CONTROLLERSMAPPED.DEVICEINSTANCE);
                    if (deviceInstance != null) equipmentNode.put("deviceInstance", deviceInstance); else equipmentNode.putNull("deviceInstance");
                    String protocol = eq.get(CONTROLLERSMAPPED.PROTOCOL);
                    equipmentNode.put("protocol", controllerCode != null ? (protocol != null ? protocol : "BACnet/IP") : "API");
                    String templateName = eq.get(EQUIPMENT.TEMPLATENAME);
                    if (templateName != null) equipmentNode.put("templateName", templateName); else equipmentNode.putNull("templateName");
                    equipmentNode.put("pointsDefined", points.size());
                    equipmentNode.put("status", engStatus);
                    equipmentNode.put("notes", "");
                    equipmentNode.set("livePoints", livePoints);
                    equipmentOut.add(equipmentNode);
                }
            }
        }

        ObjectNode siteNode = json.newObject();
        siteNode.put("id", siteRow.getId());
        siteNode.put("name", siteRow.getName());
        siteNode.put("mode", "api");
        siteNode.put("status", "editing");
        siteNode.put("nodeStatus", formLabelFromEntityStatus(siteRow.getStatus().getLiteral()));
        siteNode.put("siteType", siteRow.getSitetype() != null ? siteRow.getSitetype() : "");
        siteNode.put("timezone", siteRow.getTimezone() != null ? siteRow.getTimezone() : "");
        siteNode.put("displayLabel", siteRow.getDisplaylabel() != null ? siteRow.getDisplaylabel() : siteRow.getName());
        siteNode.put("description", siteRow.getDescription() != null ? siteRow.getDescription() : "");
        siteNode.put("engineeringNotes", siteRow.getEngineeringnotes() != null ? siteRow.getEngineeringnotes() : "");
        siteNode.put("icon", siteRow.getIcon() != null ? siteRow.getIcon() : "");
        siteNode.set("buildings", buildingsOut);

        return new HierarchyResult(siteNode, equipmentOut);
    }

    private ObjectNode pointToWorkspaceRow(String equipmentId, String equipmentName, Record p) {
        String unit = p.get(POINT.UNIT);
        String units = unit != null ? unit : "";
        String presentValue = p.get(POINT.PRESENTVALUE);
        String val = presentValue != null && !presentValue.isEmpty() ? presentValue : "—";
        String valueStr = !units.isEmpty() ? (val + " " + units).trim() : val;
        String pointCode = p.get(POINT.POINTCODE);

        ObjectNode row = json.newObject();
        row.put("id", equipmentId + "-" + (pointCode != null ? pointCode : p.get(POINT.ID)));
        row.put("databasePointId", p.get(POINT.ID));
        row.put("equipmentId", equipmentId);
        row.put("equipmentName", equipmentName);
        row.put("pointId", pointCode);
        row.put("pointKey", pointCode);
        row.put("pointDescription", p.get(POINT.POINTNAME));
        row.put("pointName", p.get(POINT.POINTNAME));
        row.put("pointReferenceId", pointCode);
        row.put("value", valueStr);
        row.put("units", units);
        row.put("status", "ACTIVE".equals(p.get(POINT.STATUS).getLiteral()) ? "OK" : "Warn");
        row.put("writable", Boolean.TRUE.equals(p.get(POINT.WRITABLE)));
        return row;
    }
}
